### PR TITLE
Fix tagrelative option not considered in builtin.tags

### LINKED ISSUE #1309: builtin tag picker does not respect tagrelative option
### Description

I keep my ctags inside the `.git` folder of my repos to not clutter up the source tree.

For that I generate my tags with `ctags --tag-relative -f .git/tags ...` and add `.git/tags` to the `tags` option.

By default `tagrelative` is set in neovim and indeed `nvim` correctly reads the tag file, as confirmed by running `:tselect` with the cursor on `main`.

However for `:Telescope tags`, the relative path (`../main.c`)  is taken relative to the current working directory and not relative to the directory the tags file is in, and so neither the preview nor opening the file works.

### Neovim version

```
NVIM v0.5.1
Build type: Release
LuaJIT 2.0.5
Compilation: /usr/bin/cc -D_FORTIFY_SOURCE=2 -march=x86-64 -mtune=generic -O2 -pipe -fno-plt -U_FORTIFY_SOURCE -D_FORTIFY_SOURCE=1 -DNVIM_TS_HAS_SET_MATCH_LIMIT -O2 -DNDEBUG -Wall -Wextra -pedantic -Wno-unused-parameter -Wstrict-prototypes -std=gnu99 -Wshadow -Wconversion -Wmissing-prototypes -Wimplicit-fallthrough -Wvla -fstack-protector-strong -fno-common -fdiagnostics-color=always -DINCLUDE_GENERATED_DECLARATIONS -D_GNU_SOURCE -DNVIM_MSGPACK_HAS_FLOAT32 -DNVIM_UNIBI_HAS_VAR_FROM -DMIN_LOG_LEVEL=3 -I/build/neovim/src/neovim-0.5.1/build/config -I/build/neovim/src/neovim-0.5.1/src -I/usr/include -I/build/neovim/src/neovim-0.5.1/build/src/nvim/auto -I/build/neovim/src/neovim-0.5.1/build/include
Compiled by builduser

Features: +acl +iconv +tui
See ":help feature-compile"

   system vimrc file: "$VIM/sysinit.vim"
  fall-back for $VIM: "/usr/share/nvim"

Run :checkhealth for more info
```

### Operating system and version

ArchLinux

### checkhealth telescope

```markdown
health#telescope#check
========================================================================
## Checking for required plugins
  - OK: plenary installed.
  - OK: nvim-treesitter installed.

## Checking external dependencies
  - OK: rg: found ripgrep 13.0.0
  - OK: fd: found fd 8.2.1

## ===== Installed extensions =====
```

### Steps to reproduce

```bash
mkdir .git
echo "int main() { return 0; }" > main.c
ctags --tag-relative --fields=afiKlmnsSzt -f ".git/tags" main.c
nvim -u minimal.lua
```

In nvim, run `:Telescope tags`

### Expected behavior

The `main.c` file is previewed correctly and opened once the tag is selected

### Actual behavior

The preview is empty and the wrong path is opened (`../main.c` instead of `main.c`).
2021-10-06-022649_2060x1360_scrot

### Minimal config

```Lua
vim.cmd [[set runtimepath=$VIMRUNTIME]]
vim.cmd [[set packpath=/tmp/nvim/site]]
local package_root = '/tmp/nvim/site/pack'
local install_path = package_root .. '/packer/start/packer.nvim'
local function load_plugins()
  require('packer').startup {
    {
      'wbthomason/packer.nvim',
      {
        'nvim-telescope/telescope.nvim',
        requires = {
          'nvim-lua/plenary.nvim',
          { 'nvim-telescope/telescope-fzf-native.nvim', run = 'make' },
        },
      },
      -- ADD PLUGINS THAT ARE _NECESSARY_ FOR REPRODUCING THE ISSUE
    },
    config = {
      package_root = package_root,
      compile_path = install_path .. '/plugin/packer_compiled.lua',
      display = { non_interactive = true },
    },
  }
end
_G.load_config = function()
  require('telescope').setup()
  require('telescope').load_extension('fzf')
  vim.opt.tags:prepend('.git/tags')
end
if vim.fn.isdirectory(install_path) == 0 then
  print("Installing Telescope and dependencies.")
  vim.fn.system { 'git', 'clone', '--depth=1', 'https://github.com/wbthomason/packer.nvim', install_path }
end
load_plugins()
require('packer').sync()
vim.cmd [[autocmd User PackerComplete ++once echo "Ready!" | lua load_config()]]
```

### PR DESCRIPTION
**Pull request by OscarCreator**
# Description

This makes buitin.tags respect tagrelative option. Instead of using `cat` command to show content of tags file `rg` or `grep` is used as they are able to print which file the line is from.

Fixes #1309

## Type of change

- Bug fix (non-breaking change which fixes an issue)

# How Has This Been Tested?

Create a tag which is **not** located in the directory you open nvim but in a subfolder. The tags in the tag file are then relative to the subfolder and not nvim's cwd. Run `:Telescope tags` when you have a file open in the subfolder, this should result in showing the tagpaths relative to nvim's cwd.

**Configuration**:
* Neovim version (nvim --version): 0.9.0
* Operating system and version: Ubuntu 22.04.2 LTS

# Checklist:

- [x] My code follows the style guidelines of this project (stylua)
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation (lua annotations)

### CODE REVIEW

File: lua/telescope/make_entry.lua [hunk 1]
```diff
@@ -1086,7 +1086,14 @@ function make_entry.gen_from_ctags(opts)
...
   return function(line)
-    if line == "" or line:sub(1, 1) == "!" then
+    if line == "" then
+      return nil
+    end
+    local tag_file
+    tag_file, line = string.match(line, '([^:]+):(.+)')
+
+    -- do not include tag file format
+    if line:sub(1, 1) == "!" then
```
Comment (OscarCreator) [inline lua/telescope/make_entry.lua:1096]: Should we filter this out when searching instead?

File: lua/telescope/make_entry.lua [hunk 2]
```diff
@@ -1098,6 +1105,12 @@ function make_entry.gen_from_ctags(opts)
       tag, file, lnum = string.match(line, "([^\t]+)\t([^\t]+)\t(%d+).*")
     end
 
+    -- append tag file path
+    if vim.opt.tagrelative:get() then
+      local tag_path = Path:new(tag_file):parent():make_relative(cwd)
+      file = tag_path .. "/" .. file
```
Comment (OscarCreator) [inline lua/telescope/make_entry.lua:1111]: Is a better way to append paths?

**Review by jamestrew — CHANGES_REQUESTED** [review]
Hey
I'm pretty unfamiliar with ctags so thanks for taking the initiative on this.

Things are looking pretty good, just a couple of thoughts.

File: lua/telescope/make_entry.lua (hunk 1 above)
Comment (OscarCreator) [reply lua/telescope/make_entry.lua]: Done

File: lua/telescope/make_entry.lua (hunk 2 above)
Comment (OscarCreator) [reply lua/telescope/make_entry.lua]: This should now be fixed.

**Review by jamestrew — APPROVED** [review]

**Comment by jamestrew** [comment]
thanks!

**Comment by jamestrew** [comment]
Sorry had to revert this due to some performance issues.

**Comment by OscarCreator** [comment]
Could you elaborate so I can fix it?

**Comment by Conni2461** [comment]
> Sorry had to revert this due to some performance issues.

All good, i'll try to be more active and warn you about code i wouldn't change without "more proper testing". This was my bad, i am sorry

> Could you elaborate so I can fix it?

take a look at https://github.com/nvim-telescope/telescope.nvim/issues/2627

The problem is, for tag files with 100k entries calculating the resulting path for every of these entries from scratch isn't performant. Additionally doing `rg -H -N --no-heading --color never -v '^!|^$'` and parsing it out adds more memory and costs additional cpu time.

Its possible to solve this, but its a none trivial problem to solve.
My idea always was to dynamically generate a function for each file that correctly calculates the path only based on string manipulation because that would be fast enough. But i never gotten around to doing that ...

**Pull request merged**